src/SonarScanner.MSBuild.PreProcessor/WebService/SonarWebService.cs

```diff
@@ -29,34 +29,43 @@
 using SonarScanner.MSBuild.PreProcessor.Protobuf;
 using SonarScanner.MSBuild.PreProcessor.Roslyn.Model;
 
-namespace SonarScanner.MSBuild.PreProcessor
+namespace SonarScanner.MSBuild.PreProcessor.WebService
 {
-    public sealed class SonarWebService : ISonarWebService
+    public abstract class SonarWebService : ISonarWebService
     {
         private const string OldDefaultProjectTestPattern = @"[^\\]*test[^\\]*$";
         private const string TestProjectPattern = "sonar.cs.msbuild.testProjectPattern";
+
+        protected readonly IDownloader downloader;
+        protected readonly Version serverVersion;
+        protected readonly ILogger logger;
         private readonly Uri serverUri;
-        private readonly IDownloader downloader;
-        private readonly ILogger logger;
-        private Version serverVersion;
 
-        public SonarWebService(IDownloader downloader, string server, ILogger logger)
-        {
-            Contract.ThrowIfNullOrWhitespace(server, nameof(server));
+        private bool disposed;
 
+        public Version ServerVersion => serverVersion;
+
+        protected SonarWebService(IDownloader downloader, Uri serverUri, Version serverVersion, ILogger logger)
+        {
             this.downloader = downloader ?? throw new ArgumentNullException(nameof(downloader));
+            this.serverUri = serverUri ?? throw new ArgumentNullException(nameof(serverUri));
+            this.serverVersion = serverVersion ?? throw new ArgumentNullException(nameof(serverVersion));
             this.logger = logger ?? throw new ArgumentNullException(nameof(logger));
 
-            // If the baseUri has relative parts (like "/api"), then the relative part must be terminated with a slash, (like "/api/"),
-            // if the relative part of baseUri is to be preserved in the constructed Uri.
-            // See: https://learn.microsoft.com/en-us/dotnet/api/system.uri.-ctor?view=net-7.0
-            serverUri = new(server.EndsWith("/") ? server : server + "/");
+            if (!serverUri.ToString().EndsWith("/"))
+            {
+                throw new ArgumentException($"{nameof(serverUri)} should always end with '/'", nameof(serverUri));
+            }
         }
 
+        public abstract Task<bool> IsServerLicenseValid();
+
+        public abstract bool IsSonarCloud();
```

Comment:
Abstract members should be above constructor and properties. And don't need empty line between them